server.go

```diff
@@ -96,50 +112,94 @@ func (srv *Server) Handler(channel string) http.HandlerFunc {
+				}
+				return false // if this happens, we'll end the handler early because something's clearly broken
+			}
+			flusher.Flush()
+			return true
+		}
+
+		var readMainCh <-chan eventOrComment = eventCh
+		var readBatchCh <-chan Event
+		closedNormally := false
+
+	ReadLoop:
 		for {
 			select {
 			case <-notifier.CloseNotify():
-				srv.unregister <- sub
-				return
+				break ReadLoop
 			case <-maxConnTimeCh: // if MaxConnTime was not set, this is a nil channel and has no effect on the select
-				srv.unregister <- sub // we treat this the same as if the client closed the connection
-				return
-			case ev, ok := <-sub.out:
+				break ReadLoop
+			case ev, ok := <-readMainCh:
 				if !ok {
-					return
+					closedNormally = true
+					break ReadLoop
 				}
-				if err := enc.Encode(ev); err != nil {
-					srv.unregister <- sub
-					if srv.Logger != nil {
-						srv.Logger.Println(err)
-					}
-					return
+				if batch, ok := ev.(eventBatch); ok {
+					readBatchCh = batch.events
+					readMainCh = nil
+				} else if !writeEventOrComment(ev) {
+					break ReadLoop
+				}
+			case ev, ok := <-readBatchCh:
+				if !ok { // end of batch
+					readBatchCh = nil
+					readMainCh = eventCh
+				} else if !writeEventOrComment(ev) {
+					break ReadLoop
 				}
-				flusher.Flush()
 			}
 		}
+		if !closedNormally {
+			srv.unsubs <- sub // the server didn't tell us to close, so we must tell it that we're closing
+		}
 	}
 }
 
-// Register registers the repository to be used for the specified channel.
+// Register registers a Repository to be used for the specified channel. The Repository will be used to
+// determine whether new subscribers should receive data that was generated before they subscribed.
+//
+// Channels do not have to be registered unless you want to specify a Repository. An unregistered channel can
+// still be subscribed to with Handler, and published to with Publish.
 func (srv *Server) Register(channel string, repo Repository) {
 	srv.registrations <- &registration{
 		channel:    channel,
 		repository: repo,
 	}
 }
 
+// Unregister removes a channel registration that was created by Register. If forceDisconnect is true, it also
+// causes all currently active handlers for that channel to close their connections. If forceDisconnect is false,
+// those connections will remain open until closed by their clients but will not receive any more events.
+//
+// This will not prevent creating new channel subscriptions for the same channel with Handler, or publishing
+// events to that channel with Publish. It is the caller's responsibility to avoid using channels that are no
+// longer supposed to be used.
+func (srv *Server) Unregister(channel string, forceDisconnect bool) {
+	srv.unregistrations <- &unregistration{
+		channel:         channel,
+		forceDisconnect: forceDisconnect,
+	}
+}
+
 // Publish publishes an event to one or more channels.
 func (srv *Server) Publish(channels []string, ev Event) {
 	srv.pub <- &outbound{
```

Comment:
Would it make sense for `Publish` to defer to `PublishWithAcknowledgement` and just swallow the acknowledgement?

Reply:
I just didn't want to add overhead to the typical use case where the caller doesn't care. Since the logic here is very simple, I don't see the advantage of unifying the two.